--- glpk/src/glplpx01.c.orig	2009-12-21 10:00:00.000000000 +0100
+++ glpk/src/glplpx01.c	2009-12-25 14:36:19.000000000 +0100
@@ -520,7 +520,7 @@
       ret = glp_simplex(lp, &parm);
       switch (ret)
       {  case 0:           ret = LPX_E_OK;      break;
-         case GLP_EBADB:
+         case GLP_EBADB:   ret = LPX_E_BADB;    break;
          case GLP_ESING:
          case GLP_ECOND:
          case GLP_EBOUND:  ret = LPX_E_FAULT;   break;
